src/Runner.Worker/ActionManager.cs

```diff
@@ -571,34 +625,55 @@ private async Task DownloadRepositoryActionAsync(IExecutionContext executionCont
                                 }
 
                                 httpClient.DefaultRequestHeaders.UserAgent.AddRange(HostContext.UserAgents);
-                                using (var result = await httpClient.GetStreamAsync(archiveLink))
+                                using (var response = await httpClient.GetAsync(link))
                                 {
-                                    await result.CopyToAsync(fs, _defaultCopyBufferSize, actionDownloadCancellation.Token);
-                                    await fs.FlushAsync(actionDownloadCancellation.Token);
-
-                                    // download succeed, break out the retry loop.
-                                    break;
+                                    if (response.IsSuccessStatusCode)
+                                    {
+                                        using (var result = await httpClient.GetStreamAsync(link))
+                                        {
+                                            await result.CopyToAsync(fs, _defaultCopyBufferSize, actionDownloadCancellation.Token);
+                                            await fs.FlushAsync(actionDownloadCancellation.Token);
+
+                                            // download succeed, break out the retry loop.
+                                            break;
+                                        }
+                                    }
+                                    else if (response.StatusCode == HttpStatusCode.NotFound)
```

Comment:
are we sure 404 means the repo/ref doesn't exist? will it happen for a bad cache on dotcom? ex, we were hitting 404 when query commit

Reply:
You mean if the user put the wrong ref in, would we get a 404, and is it appropriate to say the action doesn't exist? I think so in that case, though we could add the ref to the error message to make that clearer.  

404 was the best way I could think of to detect this.  We don't want to retry in this case, since that would slow down every run where the action actually is in the fallback URL. Open to other ideas though.